### PR TITLE
Advanced Pandas slides

### LINKED ISSUE #20: Slides: Advanced Pandas
Sources:

Currently, the slides on pandas are somewhat spread out. Advanced material for pandas can be drawn from:

* Intro class:
  * [Summarizing Grouped Data](https://github.com/uc-python/intro-python-datasci/blob/master/notebooks/09-Summarizing-Grouped-Data.ipynb)
  * [Joining Data](https://github.com/uc-python/intro-python-datasci/blob/master/notebooks/10-Joining-Data.ipynb)
  * [Exporting Data](https://github.com/uc-python/intro-python-datasci/blob/master/notebooks/11-Exporting-Data.ipynb)
  * [Visualizing Data](https://github.com/uc-python/intro-python-datasci/blob/master/notebooks/11-Exporting-Data.ipynb)
* BANA 6043:
  * [Lesson 3c](https://github.com/bradleyboehmke/uc-bana-6043/blob/main/book/03-module/lesson-3c.ipynb)
  * [Lesson 4a](https://github.com/bradleyboehmke/uc-bana-6043/blob/main/book/04-module/lesson-4a.ipynb)
  * [Lesson 4b](https://github.com/bradleyboehmke/uc-bana-6043/blob/main/book/04-module/lesson-4b.ipynb)
  * [Lesson 5a](https://github.com/bradleyboehmke/uc-bana-6043/blob/main/book/05-module/lesson-5a.ipynb)

### PR DESCRIPTION
**Pull request by cunningjames**
Closes #20.

Most of this is either new or substantially revised to fit the 45 minute time limit. It's possible some of this code might be confusing ... happy to revise if you think it necessary.

### CODE REVIEW

File: notebooks/11-Advanced-Pandas.ipynb [hunk 1]
```diff
@@ -0,0 +1,2864 @@
...
+    },
+    "tags": []
+   },
+   "source": [
+    "In pandas, a distinction that will probably trip everyone up at least once is that between **views** (looking at part of an existing object) and **copies** (making a new copy of an object in memory). This is a bit abstract, and even [the pandas docs](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy) state the following:\n",
+    "\n",
+    "> ...it’s very hard to predict whether [an operation] will return a view or a copy.\n",
+    "\n",
+    "I recommend reading the following section:\n",
+    "https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
```
Comment (augustopher) [inline notebooks/11-Advanced-Pandas.ipynb]: I would make this a hyperlink to be cleaner, but not a big deal.

File: notebooks/11-Advanced-Pandas.ipynb [hunk 2]
```diff
@@ -0,0 +1,2864 @@
...
+   "cell_type": "markdown",
+   "id": "029a4fee-6c8b-4d49-928d-5d8bdf46f9e5",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    },
+    "tags": []
+   },
+   "source": [
+    "I want you to get a sense for how this is possible, but if you don't follow all of the Python code here that's fine. Just be aware of the capability so you can research it further."
```
Comment (augustopher) [inline notebooks/11-Advanced-Pandas.ipynb]: Maybe add **`Note:`** here, to make sure this call-out is very obvious.

File: notebooks/11-Advanced-Pandas.ipynb [hunk 3]
```diff
@@ -0,0 +1,2864 @@
...
+   "cell_type": "markdown",
+   "id": "28281b46-92d8-425e-be31-c2c193110675",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    },
+    "tags": []
+   },
+   "source": [
+    "Now we need to split the data frame up into 4 portions, one ofr each core. For this we'll use the `array_split()` function from NumPy, which does exactly what it says, in this case returning a list of 4 data frames:"
```
Comment (augustopher) [inline notebooks/11-Advanced-Pandas.ipynb]:
```suggestion
    "Now we need to split the data frame up into 4 portions, one for each core. For this we'll use the `array_split()` function from NumPy, which does exactly what it says, in this case returning a list of 4 data frames:"
```

File: notebooks/11-Advanced-Pandas.ipynb [hunk 4]
```diff
@@ -0,0 +1,2864 @@
...
+   "cell_type": "markdown",
+   "id": "932e9eec-e941-455b-9824-42d1827e55bd",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    },
+    "tags": []
+   },
+   "source": [
+    "## Joining Data Frames"
```
Comment (augustopher) [inline notebooks/11-Advanced-Pandas.ipynb]: Should we move the joining section earlier? I would think we want to move from more common to less common use cases (with some flexibility for naturally building concepts), and joining data frames should be more relevant/common than parallelization.

File: notebooks/11-Advanced-Pandas.ipynb [hunk 5]
```diff
@@ -0,0 +1,2864 @@
...
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    },
+    "tags": []
+   },
+   "source": [
+    "We have several different options:"
+   ]
+  },
```
Comment (augustopher) [inline notebooks/11-Advanced-Pandas.ipynb]: I feel like this fragment is a little unnecessary, at least hanging at the bottom of this slide.

File: notebooks/11-Advanced-Pandas.ipynb [hunk 6]
```diff
@@ -0,0 +1,2864 @@
...
+   "cell_type": "markdown",
+   "id": "2f9c6bbb-cae4-4bfc-a1e7-b5f401447cd7",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    },
+    "tags": []
+   },
+   "source": [
+    "#### Left Outer Join"
```
Comment (augustopher) [inline notebooks/11-Advanced-Pandas.ipynb]: I would up this to `h3`.

```suggestion
    "### Left Outer Join"
```

File: notebooks/11-Advanced-Pandas.ipynb [hunk 7]
```diff
@@ -0,0 +1,2864 @@
...
+   "cell_type": "markdown",
+   "id": "b351d2eb-67a6-49b2-9f12-6b69091bcbab",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    },
+    "tags": []
+   },
+   "source": [
+    "#### Right Outer Join"
```
Comment (augustopher) [inline notebooks/11-Advanced-Pandas.ipynb]: I would up this to `h3`.

```suggestion
    "### Right Outer Join"
```

File: notebooks/11-Advanced-Pandas.ipynb [hunk 8]
```diff
@@ -0,0 +1,2864 @@
...
+   "cell_type": "markdown",
+   "id": "efc8e6bf-5ad6-49a7-9ec7-b2fb05b03cc4",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    },
+    "tags": []
+   },
+   "source": [
+    "#### Full Outer Join"
```
Comment (augustopher) [inline notebooks/11-Advanced-Pandas.ipynb]: I would up this to `h3`.

```suggestion
    "### Full Outer Join"
```

**Review by augustopher — CHANGES_REQUESTED** [review]
Just some small typos and style suggestions, with one slightly larger re-org idea.

**Comment by cunningjames** [comment]
@augustopher, I think I've addressed all your comments.

I also added a section on cross joins. It's a bit more niche but it was the only join type I was leaving out, so I figured I may as well be complete.

**Review by augustopher — APPROVED** [review]

**Pull request merged**